Make RRefContext get devices from RPC agent when creating OwnerRRef

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57443 Make RRefContext get devices from RPC agent when creating OwnerRRef**
* #57442 Replace DeviceIndexes with Devices in RRefs
* #56918 Make DataPtr extraction in CUDAFuture faster for Python values
* #57432 Ensure devices are preserved when forwarding between futures
* #57433 Avoid re-extracting DataPtrs when forwarding values between Futures

Based on the comments in https://github.com/pytorch/pytorch/pull/57355, I started looking at the callsites of `getOrCreateOwnerRRef` and `createOwnerRRef`, and noticed that many of them didn't specify the `devices` argument, which was optional and thus defaulted to `{}`, which created a CPU-only Future inside the OwnerRRef. (Such callsites were, for example, in `processPythonRemoteCall` and `processBaseScriptRemoteCall`, or `PyRRef::unpickle`, ...).

Some (or all?) of these callsites might still have worked thanks to the RRef's own handling of CUDA streams and events, however we intend to remove that in https://github.com/pytorch/pytorch/pull/57355. I think it would be a safer and more generic solution to always create OwnerRRefs with the full set of devices supported by the RPC agent, and this is in fact easy to do since the RRefContext has access to the RPC agent. This means that all OwnerRRefs, no matter how they're created, will support CUDA if the agent does. This also allows us to stop requiring to specify devices when creating a OwnerRRef by hand in Python.

Differential Revision: [D28144365](https://our.internmc.facebook.com/intern/diff/D28144365/)

lw committed:
Make RRefContext get devices from RPC agent when creating OwnerRRef

Based on the comments in https://github.com/pytorch/pytorch/pull/57355, I started looking at the callsites of `getOrCreateOwnerRRef` and `createOwnerRRef`, and noticed that many of them didn't specify the `devices` argument, which was optional and thus defaulted to `{}`, which created a CPU-only Future inside the OwnerRRef. (Such callsites were, for example, in `processPythonRemoteCall` and `processBaseScriptRemoteCall`, or `PyRRef::unpickle`, ...).

Some (or all?) of these callsites might still have worked thanks to the RRef's own handling of CUDA streams and events, however we intend to remove that in https://github.com/pytorch/pytorch/pull/57355. I think it would be a safer and more generic solution to always create OwnerRRefs with the full set of devices supported by the RPC agent, and this is in fact easy to do since the RRefContext has access to the RPC agent. This means that all OwnerRRefs, no matter how they're created, will support CUDA if the agent does. This also allows us to stop requiring to specify devices when creating a OwnerRRef by hand in Python.

Differential Revision: [D28144365](https://our.internmc.facebook.com/intern/diff/D28144365/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -368,14 +367,13 @@ c10::intrusive_ptr<OwnerRRef> RRefContext::getOrCreateOwnerRRef(
   // to another worker.
   return c10::make_intrusive<OwnerRRef>(
-      getWorkerId(), genGloballyUniqueId(), type, std::move(devices));
+      getWorkerId(), genGloballyUniqueId(), type, agent_->getDevices());
```
This means RRef will always sync CUDA streams even if users just want to use RRef for CPU values, and will always need to inspect values wrapped by the RRef? Will there be perf concerns? If so, shall we try some micro-benchmark (say use RRef to wrap a cpu-only ResNet-152)?

lw replied:
Right, that's a good point. I must admit, one part of me thinks we should just bite the bullet and always extract DataPtrs, even in CPU-only mode, and pay whatever perf price, since by doing this we would catch much more easily those silent CUDA sync correctness issues (e.g., all the ones I've fixed in #57432). I'm suggesting we do it not just for RRefs but for all Futures in general. In other words, correctness outweighs performance. However this only matters when the code contains bugs: in case of correct code, we'd pay this price for nothing, and that's a bummer.

I will try to do some benchmarks of this so we can get a feel of how much perf we'd be losing. I also think TorchScript here could help substantially (as inspecting an IValue is cheaper than inspecting a Python object), hence I wonder if we could have TorchScript become our "silver bullet" for performance or if it has too many feature gaps to be widely usable in RPC?

mrshenli replied:
> I wonder if we could have TorchScript become our "silver bullet" for performance

This might be hard. Based on feedback from users, some of them would like to avoid the type/API restrictions in TorchScript, even if that means they will need to take the perf hit.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -227,14 +227,9 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
               py::arg("value"),
-              py::arg("type_hint") = py::none(),
-              py::kw_only(),
-              py::arg("devices") = std::vector<c10::Device>())
```
update the argument doc?

mrshenli reviewed (approved):
Discussed with @lw offline on this. We are going to let RRefs get devices from TensorPipeAgent and always inspect values. If there are use-cases that require skipping inspect (to avoid additional pickling overhead), we can add a flag then, which won't be BC-breaking.

lw committed:
Update on "Make RRefContext get devices from RPC agent when creating OwnerRRef"


Based on the comments in https://github.com/pytorch/pytorch/pull/57355, I started looking at the callsites of `getOrCreateOwnerRRef` and `createOwnerRRef`, and noticed that many of them didn't specify the `devices` argument, which was optional and thus defaulted to `{}`, which created a CPU-only Future inside the OwnerRRef. (Such callsites were, for example, in `processPythonRemoteCall` and `processBaseScriptRemoteCall`, or `PyRRef::unpickle`, ...).

Some (or all?) of these callsites might still have worked thanks to the RRef's own handling of CUDA streams and events, however we intend to remove that in https://github.com/pytorch/pytorch/pull/57355. I think it would be a safer and more generic solution to always create OwnerRRefs with the full set of devices supported by the RPC agent, and this is in fact easy to do since the RRefContext has access to the RPC agent. This means that all OwnerRRefs, no matter how they're created, will support CUDA if the agent does. This also allows us to stop requiring to specify devices when creating a OwnerRRef by hand in Python.

Differential Revision: [D28144365](https://our.internmc.facebook.com/intern/diff/D28144365/)

[ghstack-poisoned]
